lambda/translcore.ml

```diff
@@ -88,42 +88,52 @@ type binding =
   | Bind_value of value_binding list
   | Bind_module of Ident.t * string option loc * module_presence * module_expr
 
-let rec push_defaults loc bindings cases partial =
+let wrap_bindings bindings exp =
+  List.fold_left
+    (fun exp binds ->
+      {exp with exp_desc =
+       match binds with
+       | Bind_value binds -> Texp_let(Nonrecursive, binds, exp)
+       | Bind_module (id, name, pres, mexpr) ->
+           Texp_letmodule (Some id, name, pres, mexpr, exp)})
+    exp bindings
+
+let trivial_pat pat =
+  match pat.pat_desc with
+    Tpat_var _
+  | Tpat_any -> true
+  | Tpat_construct (_, cd, [], _) ->
+      not cd.cstr_generalized && cd.cstr_consts = 1 && cd.cstr_nonconsts = 0
+  | _ -> false
+
+let rec push_defaults loc bindings cases partial need_pat =
   match cases with
     [{c_lhs=pat; c_guard=None;
       c_rhs={exp_desc = Texp_function { arg_label; param; cases; partial; } }
-        as exp}] ->
-      let cases = push_defaults exp.exp_loc bindings cases partial in
+        as exp}] when bindings = [] || trivial_pat pat ->
+      let cases = push_defaults exp.exp_loc bindings cases partial false in
```

Comment:
I don't understand why need_pat becomes false here

Reply:
`need_pat` is true when the `bindings` depend on the `c_lhs` part of cases.
This becomes true when we had something to `bindings` (as it depends on the pattern we have just seen), bu becomes false after we switch to new `cases`, as is the case here.

Reply:
Thanks for the explanation. It took me a little while to understand this logic, but it makes sense now.